src/generic/FreeAssAlgebraGroebner.jl

```diff
@@ -0,0 +1,626 @@
+###############################################################################
+#
+#   FreeAssAlgebraGroebner.jl : free associative algebra R<x1,...,xn> Groebner basis
+#
+###############################################################################
+
+export groebner_basis
+
+export interreduce!
+
+export normal_form
+
+export normal_form_weak
+
+import DataStructures: PriorityQueue, enqueue!, dequeue!
+
+const groebner_debug_level = 0
+const Monomial = Vector{Int}
+
+abstract type Obstruction{T} end
+"""
+represents the overlap of the leading term of two polynomials
```

Comment:
Could you also explain *how* it represents the overlap? i.e. what is first/second poly? what are the four entries of `pre_and_suffixes`, etc.?